File: haven/calls.py
"""Voice and video calls, layered on the exact same end-to-end encrypted
1:1 channel as text and attachments — call signaling (offer/answer/
reject/end) and every audio/video frame are just messages with
kind="call", riding whatever transport (direct LAN or relay) is already
connected to that contact. No WebRTC, no ICE/STUN/TURN, no separate
media key exchange: this reuses network.NetworkManager's existing
ratchet session, so a call is exactly as private as a text message.

The honest tradeoff for building it this way instead of real-time
SRTP-over-UDP: every audio/video chunk pays the cost of a full encrypted
message (ratchet step, JSON framing, a TCP write) rather than a
lightweight unencrypted-transport-but-content-encrypted RTP packet, and
delivery is in-order/reliable rather than "drop late packets" the way
real-time media transports prefer. For a chat app between friends over a
home relay or LAN this is perfectly usable (roughly walkie-talkie/early
Skype quality expectations), not enterprise-grade telephony — see
ROADMAP.md.

Capture/playback uses `sounddevice` (audio) and `opencv-python` (video),
both optional: a missing library or a denied OS microphone/camera
permission is reported as a call error rather than crashing the app.
"""

from __future__ import annotations

import base64
import json
import os
import threading
import time

try:
    import numpy as np
    import sounddevice as sd

    HAS_AUDIO = True
except ImportError:
    HAS_AUDIO = False

try:
    import cv2

    HAS_VIDEO = True
except ImportError:
    HAS_VIDEO = False

AUDIO_SAMPLE_RATE = 16000
AUDIO_BLOCK_SAMPLES = 1600  # 100ms per chunk @ 16kHz — see module docstring on the latency tradeoff
VIDEO_FPS = 7
VIDEO_SIZE = (320, 240)
VIDEO_JPEG_QUALITY = 50


class CallManager:
    def __init__(self, net, identity, username: str):
        self.net = net
        self.identity = identity
        self.username = username

        self.on_incoming_call = None  # callback(fingerprint, call_id, has_video)
        self.on_call_state = None  # callback(fingerprint, state) state in ringing_out/active/ended/rejected/error
        self.on_call_error = None  # callback(fingerprint, message)
        self.on_remote_video_frame = None  # callback(fingerprint, jpeg_bytes)
        self.on_local_video_frame = None  # callback(fingerprint, jpeg_bytes) — for a self-view preview
        self.on_remote_audio_chunk = None  # callback(fingerprint, pcm_bytes) — additive, e.g. live captioning

        self.calls: dict[str, dict] = {}  # fingerprint -> call state

    # -- signaling ------------------------------------------------------------

    def start_call(self, fingerprint: str, video: bool = False) -> str:
        call_id = os.urandom(8).hex()
        self.calls[fingerprint] = {"call_id": call_id, "video": video, "state": "ringing_out"}
        self.net.send_text(
            fingerprint, json.dumps({"type": "call_offer", "call_id": call_id, "video": video}), kind="call"
        )
        self._notify_state(fingerprint, "ringing_out")
        return call_id

    def accept_call(self, fingerprint: str) -> None:
        call = self.calls.get(fingerprint)
        if call is None:
            return
        self.net.send_text(
            fingerprint, json.dumps({"type": "call_answer", "call_id": call["call_id"], "accepted": True}), kind="call"
        )
        call["state"] = "active"
        self._start_media(fingerprint, call["video"])
        self._notify_state(fingerprint, "active")

    def reject_call(self, fingerprint: str) -> None:
        call = self.calls.pop(fingerprint, None)
        if call is not None:
            try:
                self.net.send_text(
                    fingerprint, json.dumps({"type": "call_reject", "call_id": call["call_id"]}), kind="call"
                )
            except ConnectionError:
                pass
        self._notify_state(fingerprint, "rejected")

    def hangup(self, fingerprint: str) -> None:
        call = self.calls.pop(fingerprint, None)
        if call is not None:
            self._stop_media(call)
            try:
                self.net.send_text(fingerprint, json.dumps({"type": "call_end", "call_id": call["call_id"]}), kind="call")
            except ConnectionError:
                pass
        self._notify_state(fingerprint, "ended")

    def set_muted(self, fingerprint: str, muted: bool) -> None:
        call = self.calls.get(fingerprint)
        if call is not None:
            call["muted"] = muted

    def handle_incoming(self, fingerprint: str, text: str) -> None:
        try:
            payload = json.loads(text)
            t = payload.get("type")
            if t == "call_offer":
                self._on_offer(fingerprint, payload)
            elif t == "call_answer":
                self._on_answer(fingerprint, payload)
            elif t == "call_reject":
                self.calls.pop(fingerprint, None)
                self._notify_state(fingerprint, "rejected")
            elif t == "call_end":
                call = self.calls.pop(fingerprint, None)
                if call is not None:
                    self._stop_media(call)
                self._notify_state(fingerprint, "ended")
            elif t == "call_audio":
                self._on_audio(fingerprint, payload)
            elif t == "call_video":
                self._on_video(fingerprint, payload)
        except (KeyError, ValueError, json.JSONDecodeError):
            pass  # malformed call-control frame — drop it rather than crash

    def _on_offer(self, fingerprint: str, payload: dict) -> None:
        call_id = payload["call_id"]
        video = payload.get("video", False)
        self.calls[fingerprint] = {"call_id": call_id, "video": video, "state": "ringing_in"}
        if self.on_incoming_call:
            self.on_incoming_call(fingerprint, call_id, video)

    def _on_answer(self, fingerprint: str, payload: dict) -> None:
        call = self.calls.get(fingerprint)
        if call is None or call["call_id"] != payload["call_id"]:
            return
        if payload.get("accepted"):
            call["state"] = "active"
            self._start_media(fingerprint, call["video"])
            self._notify_state(fingerprint, "active")
        else:
            self.calls.pop(fingerprint, None)
            self._notify_state(fingerprint, "rejected")

    def _notify_state(self, fingerprint: str, state: str) -> None:
        if self.on_call_state:
            self.on_call_state(fingerprint, state)

    # -- media: pure send/receive (hardware-independent, directly testable) ---

    def send_audio_chunk(self, fingerprint: str, pcm_bytes: bytes) -> None:
        call = self.calls.get(fingerprint)
        if call is None:
            return
        seq = call.get("audio_seq", 0)
        call["audio_seq"] = seq + 1
        payload = json.dumps(
            {
                "type": "call_audio",
                "call_id": call["call_id"],
                "seq": seq,
                "pcm_b64": base64.b64encode(pcm_bytes).decode("ascii"),
            }
        )
        self.net.send_text(fingerprint, payload, kind="call")

    def _on_audio(self, fingerprint: str, payload: dict) -> None:
        call = self.calls.get(fingerprint)
        if call is None or call["call_id"] != payload["call_id"]:
            return
        pcm = base64.b64decode(payload["pcm_b64"])
        play_queue = call.get("play_queue")
        if play_queue is not None:
            play_queue.put(pcm)
        call.setdefault("received_audio_for_test", []).append(pcm)
        if self.on_remote_audio_chunk:
            self.on_remote_audio_chunk(fingerprint, pcm)

    def send_video_frame(self, fingerprint: str, jpeg_bytes: bytes) -> None:
        call = self.calls.get(fingerprint)
        if call is None:
            return
        seq = call.get("video_seq", 0)
        call["video_seq"] = seq + 1
        payload = json.dumps(
            {
                "type": "call_video",
                "call_id": call["call_id"],
                "seq": seq,
                "jpg_b64": base64.b64encode(jpeg_bytes).decode("ascii"),
            }
        )
        self.net.send_text(fingerprint, payload, kind="call")

    def _on_video(self, fingerprint: str, payload: dict) -> None:
        call = self.calls.get(fingerprint)
        if call is None or call["call_id"] != payload["call_id"]:
            return
        jpeg_bytes = base64.b64decode(payload["jpg_b64"])
        call.setdefault("received_video_for_test", []).append(jpeg_bytes)
        if self.on_remote_video_frame:
            self.on_remote_video_frame(fingerprint, jpeg_bytes)

    # -- media: real hardware capture/playback --------------------------------

    def _start_media(self, fingerprint: str, video: bool) -> None:
        call = self.calls[fingerprint]
        call["stop_event"] = threading.Event()
        call["muted"] = False
        import queue as _queue

        call["play_queue"] = _queue.Queue()

        if not HAS_AUDIO:
            self._report_error(fingerprint, "Audio libraries not installed (pip install sounddevice numpy)")
        else:
            threading.Thread(target=self._capture_audio_loop, args=(fingerprint,), daemon=True).start()
            threading.Thread(target=self._playback_audio_loop, args=(fingerprint,), daemon=True).start()

        if video:
            if not HAS_VIDEO:
                self._report_error(fingerprint, "Video library not installed (pip install opencv-python-headless)")
            else:
                threading.Thread(target=self._capture_video_loop, args=(fingerprint,), daemon=True).start()

    def _stop_media(self, call: dict) -> None:
        stop_event = call.get("stop_event")
        if stop_event is not None:
            stop_event.set()

    def _report_error(self, fingerprint: str, message: str) -> None:
        if self.on_call_error:
            self.on_call_error(fingerprint, message)

    def _capture_audio_loop(self, fingerprint: str) -> None:
        call = self.calls.get(fingerprint)
        if call is None:
            return
        try:
            with sd.InputStream(samplerate=AUDIO_SAMPLE_RATE, channels=1, dtype="int16", blocksize=AUDIO_BLOCK_SAMPLES) as stream:
                while not call["stop_event"].is_set():
                    data, _ = stream.read(AUDIO_BLOCK_SAMPLES)
                    if call.get("muted"):
                        continue
                    try:
                        self.send_audio_chunk(fingerprint, data.tobytes())
                    except ConnectionError:
                        break
        except Exception as exc:  # sounddevice raises its own PortAudioError subclasses
            self._report_error(fingerprint, f"Microphone error: {exc}")

    def _playback_audio_loop(self, fingerprint: str) -> None:
        call = self.calls.get(fingerprint)
        if call is None:
            return
        try:
            with sd.OutputStream(samplerate=AUDIO_SAMPLE_RATE, channels=1, dtype="int16", blocksize=AUDIO_BLOCK_SAMPLES) as stream:
                while not call["stop_event"].is_set():
                    try:
                        chunk = call["play_queue"].get(timeout=0.5)
                    except Exception:
                        continue
                    arr = np.frombuffer(chunk, dtype=np.int16).reshape(-1, 1)
                    stream.write(arr)
        except Exception as exc:
            self._report_error(fingerprint, f"Speaker error: {exc}")

    def _capture_video_loop(self, fingerprint: str) -> None:
        call = self.calls.get(fingerprint)
        if call is None:
            return
        cap = cv2.VideoCapture(0)
        if not cap.isOpened():
            self._report_error(fingerprint, "Could not open camera (device 0)")
            return
        try:
            interval = 1.0 / VIDEO_FPS
            while not call["stop_event"].is_set():
                ok, frame = cap.read()
                if not ok:
                    time.sleep(0.2)
                    continue
                frame = cv2.resize(frame, VIDEO_SIZE)
                ok2, buf = cv2.imencode(".jpg", frame, [cv2.IMWRITE_JPEG_QUALITY, VIDEO_JPEG_QUALITY])
                if not ok2:
                    continue
                jpeg_bytes = buf.tobytes()
                if self.on_local_video_frame:
                    self.on_local_video_frame(fingerprint, jpeg_bytes)
                try:
                    self.send_video_frame(fingerprint, jpeg_bytes)
                except ConnectionError:
                    break
                time.sleep(interval)
        except Exception as exc:
            self._report_error(fingerprint, f"Camera error: {exc}")
        finally:
            cap.release()

File: webapp/static/js/calls.js
/**
 * Voice and video calls — the browser port of haven/calls.py. Same
 * deliberate design choice as desktop: no WebRTC, no ICE/STUN/TURN, no
 * separate media key exchange. Call signaling (offer/answer/reject/end)
 * and every audio/video chunk are just messages with kind="call" riding
 * the exact same end-to-end encrypted 1:1 ratchet session as text — so a
 * call is exactly as private as a text message, and a web client can
 * call a desktop client (or vice versa) since the wire format matches
 * exactly: 16kHz mono 16-bit PCM audio chunks, JPEG video frames.
 *
 * The honest tradeoff (see calls.py's own docstring) is real: every
 * audio/video chunk pays for a full encrypted message rather than a
 * lightweight unencrypted-transport RTP packet, and delivery is
 * in-order/reliable rather than "drop late packets" the way real-time
 * media transports prefer. Walkie-talkie/early-Skype quality, not
 * enterprise telephony — that's an accepted, documented tradeoff, not
 * an oversight.
 */

const HavenCalls = (() => {
  "use strict";
  const H = Haven;

  const AUDIO_SAMPLE_RATE = 16000;
  const AUDIO_BLOCK_SAMPLES = 1600; // 100ms @ 16kHz, matches calls.py exactly
  const VIDEO_FPS = 7;
  const VIDEO_SIZE = { width: 320, height: 240 };
  const VIDEO_JPEG_QUALITY = 0.5;

  function bytesToBase64(bytes) {
    let bin = "";
    const chunkSize = 0x8000;
    for (let i = 0; i < bytes.length; i += chunkSize) bin += String.fromCharCode.apply(null, bytes.subarray(i, i + chunkSize));
    return btoa(bin);
  }
  function base64ToBytes(b64) {
    const bin = atob(b64);
    const out = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
    return out;
  }

  class CallManager {
    constructor(net, identity, username) {
      this.net = net;
      this.identity = identity;
      this.username = username;

      this.onIncomingCall = null; // (fingerprint, callId, hasVideo) => void
      this.onCallState = null; // (fingerprint, state) state in ringing_out/active/ended/rejected/error
      this.onCallError = null; // (fingerprint, message) => void
      this.onRemoteVideoFrame = null; // (fingerprint, blobUrl) => void
      this.onLocalVideoFrame = null; // (fingerprint, blobUrl) => void — self-view preview
      this.onAudioChunk = null; // (fingerprint, int16Samples) => void — raw incoming PCM, for live translation (see translation.js)

      this.calls = new Map(); // fingerprint -> call state
    }

    // -- signaling ------------------------------------------------------------

    async startCall(fingerprint, video = false) {
      const callId = H.bytesToHex(crypto.getRandomValues(new Uint8Array(8)));
      this.calls.set(fingerprint, { callId, video, state: "ringing_out" });
      await this.net.sendText(fingerprint, JSON.stringify({ type: "call_offer", call_id: callId, video }), "call");
      this._notifyState(fingerprint, "ringing_out");
      return callId;
    }

    async acceptCall(fingerprint) {
      const call = this.calls.get(fingerprint);
      if (!call) return;
      await this.net.sendText(fingerprint, JSON.stringify({ type: "call_answer", call_id: call.callId, accepted: true }), "call");
      call.state = "active";
      // The call UI (hang up/mute) must not wait on getUserMedia's
      // permission prompt — that can take seconds, be ignored, or never
      // resolve at all, and the call is already connected from a
      // signaling standpoint regardless of mic/camera permission state.
      this._notifyState(fingerprint, "active");
      this._startMedia(fingerprint, call.video).catch((e) => this._reportError(fingerprint, "Media error: " + e.message));
    }

    async rejectCall(fingerprint) {
      const call = this.calls.get(fingerprint);
      this.calls.delete(fingerprint);
      if (call) {
        try {
          await this.net.sendText(fingerprint, JSON.stringify({ type: "call_reject", call_id: call.callId }), "call");
        } catch (e) {
          /* best-effort */
        }
      }
      this._notifyState(fingerprint, "rejected");
    }

    async hangup(fingerprint) {
      const call = this.calls.get(fingerprint);
      this.calls.delete(fingerprint);
      if (call) {
        this._stopMedia(call);
        try {
          await this.net.sendText(fingerprint, JSON.stringify({ type: "call_end", call_id: call.callId }), "call");
        } catch (e) {
          /* best-effort */
        }
      }
      this._notifyState(fingerprint, "ended");
    }

    setMuted(fingerprint, muted) {
      const call = this.calls.get(fingerprint);
      if (call) call.muted = muted;
    }

    async handleIncoming(fingerprint, text) {
      let payload;
      try {
        payload = JSON.parse(text);
      } catch (e) {
        return;
      }
      try {
        const t = payload.type;
        if (t === "call_offer") await this._onOffer(fingerprint, payload);
        else if (t === "call_answer") await this._onAnswer(fingerprint, payload);
        else if (t === "call_reject") {
          this.calls.delete(fingerprint);
          this._notifyState(fingerprint, "rejected");
        } else if (t === "call_end") {
          const call = this.calls.get(fingerprint);
          this.calls.delete(fingerprint);
          if (call) this._stopMedia(call);
          this._notifyState(fingerprint, "ended");
        } else if (t === "call_audio") this._onAudio(fingerprint, payload);
        else if (t === "call_video") this._onVideo(fingerprint, payload);
      } catch (e) {
        console.error("malformed call control frame, dropped:", e);
      }
    }

    async _onOffer(fingerprint, payload) {
      const video = !!payload.video;
      this.calls.set(fingerprint, { callId: payload.call_id, video, state: "ringing_in" });
      if (this.onIncomingCall) this.onIncomingCall(fingerprint, payload.call_id, video);
    }

    async _onAnswer(fingerprint, payload) {
      const call = this.calls.get(fingerprint);
      if (!call || call.callId !== payload.call_id) return;
      if (payload.accepted) {
        call.state = "active";
        this._notifyState(fingerprint, "active");
        this._startMedia(fingerprint, call.video).catch((e) => this._reportError(fingerprint, "Media error: " + e.message));
      } else {
        this.calls.delete(fingerprint);
        this._notifyState(fingerprint, "rejected");
      }
    }

    _notifyState(fingerprint, state) {
      if (this.onCallState) this.onCallState(fingerprint, state);
    }

    // -- media: pure send/receive (hardware-independent, directly testable) ---

    async sendAudioChunk(fingerprint, pcmBytes) {
      const call = this.calls.get(fingerprint);
      if (!call) return;
      const seq = call.audioSeq || 0;
      call.audioSeq = seq + 1;
      const payload = JSON.stringify({ type: "call_audio", call_id: call.callId, seq, pcm_b64: bytesToBase64(pcmBytes) });
      await this.net.sendText(fingerprint, payload, "call");
    }

    _onAudio(fingerprint, payload) {
      const call = this.calls.get(fingerprint);
      if (!call || call.callId !== payload.call_id) return;
      const pcm = base64ToBytes(payload.pcm_b64);
      const int16 = new Int16Array(pcm.buffer, pcm.byteOffset, pcm.length / 2);
      if (call.playChunk) call.playChunk(int16);
      if (this.onAudioChunk) this.onAudioChunk(fingerprint, int16);
    }

    async sendVideoFrame(fingerprint, jpegBytes) {
      const call = this.calls.get(fingerprint);
      if (!call) return;
      const seq = call.videoSeq || 0;
      call.videoSeq = seq + 1;
      const payload = JSON.stringify({ type: "call_video", call_id: call.callId, seq, jpg_b64: bytesToBase64(jpegBytes) });
      await this.net.sendText(fingerprint, payload, "call");
    }

    _onVideo(fingerprint, payload) {
      const call = this.calls.get(fingerprint);
      if (!call || call.callId !== payload.call_id) return;
      const jpegBytes = base64ToBytes(payload.jpg_b64);
      const url = URL.createObjectURL(new Blob([jpegBytes], { type: "image/jpeg" }));
      if (this.onRemoteVideoFrame) this.onRemoteVideoFrame(fingerprint, url);
    }

    // -- media: real hardware capture/playback (getUserMedia + Web Audio) -----

    async _startMedia(fingerprint, video) {
      const call = this.calls.get(fingerprint);
      if (!call) return; // hung up already, before this (unawaited) setup got a chance to run
      call.muted = false;
      call.stopped = false;

      // Playback: a small scheduled-queue player so back-to-back 100ms PCM
      // chunks play smoothly instead of overlapping or gapping.
      const playbackCtx = new AudioContext({ sampleRate: AUDIO_SAMPLE_RATE });
      let nextPlayTime = 0;
      call.playbackCtx = playbackCtx;
      call.playChunk = (int16) => {
        const float32 = new Float32Array(int16.length);
        for (let i = 0; i < int16.length; i++) float32[i] = int16[i] / 0x8000;
        const buffer = playbackCtx.createBuffer(1, float32.length, AUDIO_SAMPLE_RATE);
        buffer.copyToChannel(float32, 0);
        const src = playbackCtx.createBufferSource();
        src.buffer = buffer;
        src.connect(playbackCtx.destination);
        const startAt = Math.max(playbackCtx.currentTime, nextPlayTime);
        src.start(startAt);
        nextPlayTime = startAt + buffer.duration;
      };

      try {
        const micStream = await navigator.mediaDevices.getUserMedia({ audio: true });
        const captureCtx = new AudioContext({ sampleRate: AUDIO_SAMPLE_RATE });
        const source = captureCtx.createMediaStreamSource(micStream);
        const processor = captureCtx.createScriptProcessor(AUDIO_BLOCK_SAMPLES, 1, 1);
        const silentGain = captureCtx.createGain();
        silentGain.gain.value = 0; // ScriptProcessorNode needs a path to destination to keep firing, but we don't want to hear ourselves
        processor.onaudioprocess = (e) => {
          if (call.stopped || call.muted) return;
          const input = e.inputBuffer.getChannelData(0);
          const pcm16 = new Int16Array(input.length);
          for (let i = 0; i < input.length; i++) {
            const s = Math.max(-1, Math.min(1, input[i]));
            pcm16[i] = s < 0 ? s * 0x8000 : s * 0x7fff;
          }
          this.sendAudioChunk(fingerprint, new Uint8Array(pcm16.buffer)).catch((e2) => console.error("send audio failed:", e2));
        };
        source.connect(processor);
        processor.connect(silentGain);
        silentGain.connect(captureCtx.destination);
        call.micStream = micStream;
        call.captureCtx = captureCtx;
        call.processor = processor;
      } catch (e) {
        this._reportError(fingerprint, "Microphone error: " + e.message);
      }

      if (video) {
        try {
          const videoStream = await navigator.mediaDevices.getUserMedia({ video: VIDEO_SIZE });
          const videoEl = document.createElement("video");
          videoEl.srcObject = videoStream;
          videoEl.muted = true;
          videoEl.playsInline = true;
          await videoEl.play();
          const canvas = document.createElement("canvas");
          canvas.width = VIDEO_SIZE.width;
          canvas.height = VIDEO_SIZE.height;
          const ctx2d = canvas.getContext("2d");
          call.videoStream = videoStream;
          call.videoTimer = setInterval(() => {
            if (call.stopped) return;
            ctx2d.drawImage(videoEl, 0, 0, canvas.width, canvas.height);
            canvas.toBlob(
              async (blob) => {
                if (!blob || call.stopped) return;
                const jpegBytes = new Uint8Array(await blob.arrayBuffer());
                if (this.onLocalVideoFrame) this.onLocalVideoFrame(fingerprint, URL.createObjectURL(blob));
                try {
                  await this.sendVideoFrame(fingerprint, jpegBytes);
                } catch (e2) {
                  console.error("send video failed:", e2);
                }
              },
              "image/jpeg",
              VIDEO_JPEG_QUALITY
            );
          }, 1000 / VIDEO_FPS);
        } catch (e) {
          this._reportError(fingerprint, "Camera error: " + e.message);
        }
      }
    }

    _stopMedia(call) {
      call.stopped = true;
      if (call.videoTimer) clearInterval(call.videoTimer);
      if (call.micStream) call.micStream.getTracks().forEach((t) => t.stop());
      if (call.videoStream) call.videoStream.getTracks().forEach((t) => t.stop());
      if (call.processor) call.processor.disconnect();
      if (call.captureCtx) call.captureCtx.close().catch(() => {});
      if (call.playbackCtx) call.playbackCtx.close().catch(() => {});
    }

    _reportError(fingerprint, message) {
      if (this.onCallError) this.onCallError(fingerprint, message);
    }
  }

  return { CallManager };
})();
